velox/substrait/SubstraitToVeloxExpr.cpp

```diff
@@ -16,6 +16,128 @@
 
 #include "velox/substrait/SubstraitToVeloxExpr.h"
 #include "velox/substrait/TypeUtils.h"
+#include "velox/vector/FlatVector.h"
+
+using namespace facebook::velox;
+namespace {
+// Get values for the different supported types.
+template <typename T>
+T getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  VELOX_NYI();
+}
+
+template <>
+int8_t getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return static_cast<int8_t>(literal.i8());
+}
+
+template <>
+int16_t getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return static_cast<int16_t>(literal.i16());
+}
+
+template <>
+int32_t getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.i32();
+}
+
+template <>
+int64_t getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.i64();
+}
+
+template <>
+double getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.fp64();
+}
+
+template <>
+float getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.fp32();
+}
+
+template <>
+bool getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.boolean();
+}
+
+template <>
+uint32_t getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.i32();
+}
+
+ArrayVectorPtr makeArrayVector(const VectorPtr& elements) {
+  BufferPtr offsets = allocateOffsets(1, elements->pool());
+  BufferPtr sizes = allocateOffsets(1, elements->pool());
+  sizes->asMutable<vector_size_t>()[0] = elements->size();
+
+  return std::make_shared<ArrayVector>(
+      elements->pool(),
+      ARRAY(elements->type()),
+      nullptr,
+      1,
+      offsets,
+      sizes,
+      elements);
+}
+
+ArrayVectorPtr makeEmptyArrayVector(memory::MemoryPool* pool) {
+  BufferPtr offsets = allocateOffsets(1, pool);
+  BufferPtr sizes = allocateOffsets(1, pool);
+  return std::make_shared<ArrayVector>(
+      pool, ARRAY(UNKNOWN()), nullptr, 1, offsets, sizes, nullptr);
+}
+
+template <TypeKind kind>
+void listLiteralToFlatVector(
+    const ::substrait::Expression::Literal& listLiteral,
+    const VectorPtr& vector,
+    vector_size_t& index) {
+  using T = typename TypeTraits<kind>::NativeType;
+  auto flatVector = vector->as<FlatVector<T>>();
+  auto typeCase = listLiteral.literal_type_case();
+  if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kList) {
```

Comment:
listLiteralToFlatVector is only called from constructFlatVector which asserts that type is a primitive type. Hence, it is not possible for typeCase to be kList, is it? If that's not possible, let's remove this code.

Reply:
Actually this code is necessary. For nested-array. The listLiteral's typeCase can be kList.  
"VELOX_CHECK(type->isPrimitiveType());" This type means the type of every basic element.
For example：
array[[1,3], [2,4]]
listLiteral in constructFlatVector will be "[ [ 1,3], [2,4 ] ]". Because he stores the entire nested-array.
type will be int.  Because he represents the type of each innermost element.